contracts/rate_oracles/CompoundBorrowRateOracle.sol

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity =0.8.9;
+
+import "../interfaces/rate_oracles/ICompoundRateOracle.sol";
+import "../interfaces/compound/ICToken.sol";
+import "./BaseRateOracle.sol";
+import "../utils/ExponentialNoError.sol";
+
+contract CompoundBorrowRateOracle is
+    BaseRateOracle,
+    ICompoundRateOracle,
+    ExponentialNoError
+{
+    /// @inheritdoc ICompoundRateOracle
+    ICToken public immutable override ctoken;
+
+    /// @inheritdoc ICompoundRateOracle
+    uint8 public immutable override decimals;
+
+    uint8 public constant override UNDERLYING_YIELD_BEARING_PROTOCOL_ID = 6; // id of compound is 6
+
+    // Maximum borrow rate that can ever be applied (.0005% / block)
+    // https://github.com/compound-finance/compound-protocol/blob/a3214f67b73310d547e00fc578e8355911c9d376/contracts/CTokenInterfaces.sol#L31
+    uint256 internal constant BORROW_RATE_MAX_MANTISSA = 0.0005e16;
+
+    constructor(
+        ICToken _ctoken,
+        bool ethPool,
+        IERC20Minimal underlying,
+        uint8 _decimals,
+        uint32[] memory _times,
+        uint256[] memory _results
+    ) BaseRateOracle(underlying) {
+        ctoken = _ctoken;
+        require(
+            ethPool || ctoken.underlying() == address(underlying),
+            "Tokens do not match"
+        );
+        // Check that underlying was set in BaseRateOracle
+        require(address(underlying) != address(0), "underlying must exist");
+        decimals = _decimals;
+
+        _populateInitialObservations(_times, _results);
+    }
+
+    /// @inheritdoc BaseRateOracle
+    // Follows the accrueInterest() logic from Compound's CToken to compute latest rate
+    // https://github.com/compound-finance/compound-protocol/blob/a3214f67b73310d547e00fc578e8355911c9d376/contracts/CToken.sol#L327
+    function getLastUpdatedRate()
+        public
+        view
+        override
+        returns (uint32 timestamp, uint256 resultRay)
+    {
+        uint256 borrowRateMantissa = ctoken.borrowRatePerBlock();
+        require(
+            borrowRateMantissa <= BORROW_RATE_MAX_MANTISSA,
+            "borrow rate is absurdly high"
+        );
+
+        uint256 blockDelta = block.number - ctoken.accrualBlockNumber();
+        // rate accrued since last index update
+        Exp memory simpleInterestFactor = mul_(
+            Exp({mantissa: borrowRateMantissa}),
+            blockDelta
+        );
+
+        uint256 borrowIndexPrior = ctoken.borrowIndex();
+        uint256 borrowIndex = mul_ScalarTruncateAddUInt(
+            simpleInterestFactor,
+            borrowIndexPrior,
+            borrowIndexPrior
+        ); // result given in wad, scale to ray
```

Comment:
can we re-confirm that the result is indeed given in wad?

e.g. looking through the supply compound rate oracle, it appears the conversion is slightly more involved as it is a function of the number of decimals of the underlying token:

https://github.com/Voltz-Protocol/voltz-core/blob/ac45489df29306b2e648dccb940d6c6eb6c2a406/contracts/rate_oracles/CompoundRateOracle.sol#L75

wdyt @CyclopsRex ?

Reply:
Yeah, I too am a little surprised if this is in wad but the maths Compound is doing does seem to imply that. Worth checking for sure.

Reply:
![Screenshot 2022-08-03 at 19 37 42](https://user-images.githubusercontent.com/60431079/182684381-1e7737af-e769-4d3c-a530-e6b450292fff.png)
taken from cDAI
![Screenshot 2022-08-03 at 19 38 50](https://user-images.githubusercontent.com/60431079/182684562-aba31d57-4283-4a16-b5c6-541ce4721688.png)
taken from cUSDC